Add `allow-mock-accuracy` for non-default `geopoint` and `geoshape`/`geotrace`

seadowg opened the pull request:
Closes #4768 
~~Blocked by #4842~~

Every kind of geo/location widget should now support `allow-mock-accuracy`. Again, the refactor here  (moving the geo code to `geo`) is nice and gradual so going through commit by commit is a good idea.

#### What has been done to verify that this works as intended?

New tests.

#### Why is this the best possible solution? Were any other approaches considered?

This takes the same approach as the last PR so not a lot to discuss here. This time most of the work is focused on moving `GeoPointMapActivity` and then adding support for the `allow-mock-accuracy` param in the same way we did for `GeoPointActivity`.

It's feeling to me like there's probably a `maps` module to be extracted at some point that contains the `MapFragment`, `MapPoint` and the providers/implementations (or possibly multiple modules for different implementations). That feels like an extra step we could take next time we're playing around in this area however.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Good to check the map appearances of `geopoint` with the different map providers - Mapbox should require special attention because it uses a different code path for grabbing the device's location. `geoshape` and `geotrace` need checked as well.

#### Do we need any specific form for testing your changes? If so, please attach one.

[The form](https://github.com/getodk/collect/files/7285696/external_gevice_geopoint.zip) used in the previous PR is a good starting point. Testing the specific support in this PR can be done by adding the map appearances to the form.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

seadowg committed:
Push previousState down to subclasses as it's use is differnet in different classes

seadowg committed:
Remove extra permission check from geo activities

seadowg committed:
Remove unneeded super class

seadowg committed:
Separate GeoPointMapActivity from CollectAbstractActivity

seadowg committed:
Decouple GeoPointMapActivity from MapPreferencesFragment

seadowg committed:
Extract interface for creating map fragments

seadowg committed:
Move maps interfaces to geo module

These could eventually move to their own `maps` module but moving
the map fragment implementations would be very cumbersome and is
probably worth doing a single dedicated piece of work.

seadowg committed:
Move activity extras constants to geo package

seadowg committed:
Move GeoPointMapActivity to geo package

seadowg committed:
Fix test imports

seadowg committed:
use fake MapFragment to test GeoPointMapActivity instead of real fragment with faked location client

seadowg committed:
Move to using ActivityScenario in GeoPointMapActivityTest

seadowg committed:
Move GeoPointMapActivityTest to geo module

seadowg committed:
Add pass-through setter for retain mock accuracy

We'll need to handle Mapbox in a different way as it doesn't use
the same location abstraction

seadowg committed:
Support retaining mock accuracy in GeoPointMapActivity

seadowg committed:
Use same location sanitizer code for mapbox as everything else

seadowg committed:
Support retaining mock accuracy in MapboxMapFragment

seadowg committed:
Add note to dagger setup

seadowg committed:
Remove dependency on CollectAbstractActivity from GeoPolyActivity

seadowg committed:
Move dialog fragment helpers to androidshared

seadowg committed:
Move GeoPolySettingsDialogFragment to geo module

seadowg committed:
Move GeoPolyActivity to geo module

seadowg committed:
Revise GeoPolyActivityTest to focus on LocationTracker rather than LocationClient

seadowg committed:
Use ActivityScenario for GeoPolyActivityTest

seadowg committed:
Move GeoPolyActivityTest to geo module

seadowg committed:
Send retain mock accuracy when starting GeoPolyActivity

seadowg committed:
Support retain mock accuracy in GeoPolyActivity

seadowg committed:
Correct test name

seadowg committed:
Extract shared test fake

seadowg committed:
Optimize imports

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormMapActivity.java:
```diff
@@ -112,8 +114,15 @@
+
         DaggerUtils.getComponent(this).inject(this);
 
+        if (!permissionsProvider.areLocationPermissionsGranted()) {
```
Before this method was in `BaseGeoMapActivity` extended by this class but also by `GeoPointMapActivity` and `GeoPolyActivity`. Are you sure it's no longer needed in those two last classes? What if a user minimizes the app, disables location permissions and reopens the app?

seadowg replied:
You're right about the minimize/deny permission case though as I'd forgotten that that kills the app and then reloads the same backstack (rather than starting from fresh). Good catch - we missed this when reviewing #4829. Right now if you perform those steps on a default `geopoint` question (`GeoPointActivity`) you get a crash caused by a `SecurityException`. 

Really we should only handle the permission at the Activity level instead of in `ActivityGeoDataRequester`. I'll make that change in this PR.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/GeoPointMapActivity.java:
```diff
@@ -70,8 +68,16 @@
     public static final String LOCATION_STATUS_VISIBILITY_KEY = "location_status_visibility";
     public static final String LOCATION_INFO_VISIBILITY_KEY = "location_info_visibility";
 
+    public static final String EXTRA_LOCATION = "gp";
```
Why you didn't move this one to `geo/Constants.kt` too like `EXTRA_READ_ONLY` and  `EXTRA_DRAGGABLE_ONLY`?

seadowg replied:
This one is only relevant to `GeoPointMapActivity`. I tried to keep it so `Constants` was only defining the shared ones.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/GeoPointMapActivity.java:
```diff
@@ -177,7 +185,7 @@ public void returnLocation() {
 
         if (result != null) {
-            setResult(RESULT_OK, new Intent().putExtra(FormEntryActivity.ANSWER_KEY, result));
+            setResult(RESULT_OK, new Intent().putExtra("value", result));
```
Now we don't use the same constant anymore what might be error prone. It would be good to move that const to a shared module if now we need to use it from different modules.

seadowg replied:
Yeah I had a comment about this on the last PR. We also assume external apps use `"value"` so it's fairly well established. I'm not sure exactly where to define it given that it's not even specific to the Collect app. Maybe we could have a helper for Activities that return data to `FormEntryActivity` that external apps could copy? Kind of like without content provider API objects...

grzesiek2010 commented on collect_app/src/main/AndroidManifest.xml:
```diff
@@ -137,9 +137,15 @@ the specific language governing permissions and limitations under the License.
-            tools:replace="android:theme"/>
-        <activity android:name=".activities.GeoPointMapActivity" />
-        <activity android:name=".activities.GeoPolyActivity" />
+            tools:replace="android:theme" />
```
What is it for?

seadowg replied:
It's need so that the build tools now which `android:theme` to use. Otherwise, there is a conflict between the `geo` and the `collect_app` manifests.

grzesiek2010 commented on geo/src/test/java/org/odk/collect/geo/GeoPolyActivityTest.java:
```diff
@@ -0,0 +1,119 @@
+        ActivityScenario.<GeoPolyActivity>launch(intent).onActivity(activity -> {
+            activity.updateRecordingMode(R.id.automatic_mode);
+            activity.startInput();
+            verify(locationTracker).start(true);
```
Why it's `true` when you passed false to `EXTRA_RETAIN_MOCK_ACCURACY`?

seadowg replied:
WAT. That's a mistake. Not sure why the test passes though as the implementation looks right. Will have to look at that.

seadowg replied:
Ah it's just that `start(true)` was called before. I'll add a `Mockito.reset` to make this test safer. This is a good example of where mocking frameworks can be dangerous in comparison to simple stateful fakes!

seadowg committed:
Make sure test won't pass incorrectly

seadowg committed:
Move PermissionsChecker and rearrange packages in androidshared

seadowg committed:
Add permission checks back into geo Activity objects

This is needed for cases where the user denies permissions in
settings while on one of these screens, forcing the app to restart
and rebuild the backstack.

seadowg commented on geo/src/main/java/org/odk/collect/geo/GeoActivityUtils.kt:
```diff
@@ -0,0 +1,22 @@
+internal object GeoActivityUtils {
+
+    @JvmStatic
+    fun requireLocationPermissions(activity: Activity) {
```
I don't like that we check this in the `Activity` and in `ActivityGeoDataRequester`. However, I don't want to get into untangling the interaction with `WaitingForDataRegistry` here to make sure that it's safe to just check this in the `Activity`.

seadowg committed:
Grant permissions in tests

seadowg committed:
Extract helper for returning single values from external app/activity

seadowg committed:
Use external app helper to pull value out in tests

seadowg committed:
Deprecate ANSWER_KEY constant

seadowg committed:
Add comment to ExternalAppUtils

seadowg committed:
Remove unused pro guard file

seadowg commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -220,6 +220,10 @@
+    /**
+     * Should use {@link org.odk.collect.android.externaldata.ExternalAppsUtils} instead
+     */
+    @Deprecated
```
I didn't want to add the work/risk of replacing current usages of `ANSWER_KEY` in this PR but also wanted to make it clear we have a better option now.

seadowg committed:
Remove unused desugaring

kkrawczyk123 commented:
@seadowg I can see a problem with the display of Google Maps when using all Geo widgets (point, trace, shape) on all of my devices. It loads properly for display points on map. Issue does not occur for Mapbox and OSM. See gif:
![ezgif com-gif-maker (7)](https://user-images.githubusercontent.com/33342488/138264653-04717055-a760-421f-a2ff-61831887b106.gif)

The issue does not occur on master.

seadowg commented:
@kkrawczyk123 I'll look into that

seadowg committed:
Use singelton MapProvider instead of new one each time

For some reason this needs to be a singleton or Google Maps ends
up just being blank. Not sure why yet.

seadowg committed:
Don't inject dependency module dependencies in application object

seadowg committed:
Add note about MapProvider needing to be a singleton

seadowg committed:
Optimize imports

seadowg commented:
@kkrawczyk123 fixed! 

@grzesiek2010 it looks like the `MapProvider` needs to be a singleton. I've noted this on the class now to hopefully avoid the same mistake being made again, but you might want to take a look and see if there's anything you think we need to address there.

kkrawczyk123 commented:
@seadowg I have tried that form https://test.getodk.cloud/#/projects/308/forms/external_device_geopoint/versions with Fake GPS app and my understanding is that fake location should be used only in the first question where mock accuracy=true but I can see that it is used also in the second question where mock accuracy=false. Is it an issue or did I misunderstand something?

seadowg commented:
@kkrawczyk123 the fake location will still be used, but its accuracy should be zero when `allow-mock-accuracy` is `false` (or absent). Is that not what you're seeing? The idea is just to remove accuracy from fake locations but allow people that have to use fake locations to be able to keep it if they trust it.

kkrawczyk123 commented:
@seadowg I can see accuracy 0 for both when allow-mock-accuracy is true and false:
|    |    | 
| --- | ---- |
| ![Screenshot (Oct 22, 2021 9_13_45 AM)](https://user-images.githubusercontent.com/33342488/138410921-d51756ac-2d36-43f4-beb4-daaa4f26ecd9.png) | ![Screenshot (Oct 22, 2021 9_13_54 AM)](https://user-images.githubusercontent.com/33342488/138410952-a25bf23c-59e7-4239-a014-568f26bd7b89.png) |

I assumed that there should be a difference but maybe I am wrong.

seadowg commented:
@kkrawczyk123 is this for all the geo widgets or just one of them? I'll check this out my side as well as it sounds like something it wrong.

seadowg commented:
@kkrawczyk123 I'm seeing the correct behaviour using mock the [GPS Connector](https://play.google.com/store/apps/details?id=de.pilablu.gpsconnector&hl=en_US&gl=US) app in simulation mode for geopoint (and maps appearances) and geotrace - I get points with 0 accuracy when `allow-mock-accuracy` is `false` and non zero when it's `true`. 

Are you potentially testing with manual placement (moving the map to somewhere and then placing a point)? This always results in zero accuracy as the location doesn't come from gps.

kkrawczyk123 commented:
@seadowg I agree that accuracy difference is visible between both questions for geopoint with map view, geotrace and geoshape widgets but my question was about geopoint without map view - exact form that was attached in PR.

seadowg commented:
@kkrawczyk123 yeah I'm using that as well and still getting accuracy in the "Location (allows mock)" question. Is that happening every time? And it doesn't happen for the other geo widgets? What app are you using to fake the location? I'm wondering if the app is maybe giving you zero accuracy - I just got a `0.01m` out of the simulation in the one I was using.

kkrawczyk123 commented:
@seadowg I can see accuracy 0 for both questions. It is happening every time for me. @mmarciniak90 Are you seeing the same thing on your devices? I am using the same app "Fake GPS" for that widget and others and there is a difference as I mentioned above it works for widgets with map view. I have set accuracy in the app to be different than 0 - I have 1 on one device and 20 on another and the same results - accuracy 0 for questions with and without allow-mock accuracy when filling a form.

seadowg commented:
@kkrawczyk123 could link me to the exact app?

kkrawczyk123 commented:
sure, https://play.google.com/store/apps/details?id=com.lexa.fakegps&hl=en&gl=US

seadowg commented:
@kkrawczyk123 I get accuracy with that app. Changing its "Accuracy" setting is reflected in Collect as well (although never exactly for some reason, but that could be the app).

mmarciniak90 commented:
> Are you seeing the same thing on your devices?

Unfortunately, I have the same result. I also use the same app to generate fake GPS.

kkrawczyk123 commented:
@seadowg That's weird. What app have you been using before? We could try with it

seadowg commented:
Ok turns out I'd got myself confused about the release state of the `allow-mock-accuracy` changes in pyxform. It's out on https://getodk.org/xlsform/ but not released for Central yet. My bad! I've shared a form created with the newest pyxform with QA for them to test with.

kkrawczyk123 commented:
**Finally tested with success!**
Verified on Androids 5.1 and 11.

Verified cases:
- form from https://test.getodk.cloud/#/projects/306/forms/external_device_geopoint
- Fake GPS app
- GPS Connector app

regression checks on:
- Geo widgets with/without map
- GeoTrace, GeoShape widgets
- placement by tapping, manual mode, automatic mode
- display submissions on map
- submissions with different statuses
- Google maps
- OSM
- Mapbox
- offline layers
- light mode, dark mode
- Audit location in the background
- screen rotation
- minimize app

mmarciniak90 commented:
Tested with success on Android 8.1, 9.0, 10.0
